### PR TITLE
feat(bigquery): implement `typeof` using SQL UDF

### PR DESCRIPTION
**Pull request by krzysztof-kwitt**
Hello,

I decided to implement these operations as a UDF so that the SQL code is still readable when the Ibiis expression contains several use of ops.Typeof.

Best regards,

### CODE REVIEW

File: ibis/backends/tests/test_generic.py [hunk 1]
```diff
@@ -760,9 +769,406 @@ def test_exists(batting, awards_players, method_name):
...
+)
+def test_typeof(backend, con, ibis_op, expected_values):
+    expr = ibis_op.typeof()
+    result = con.execute(expr)
+
+    backend_name = backend.name()
+    if backend_name in expected_values:
+        assert result == expected_values[backend_name]
+    else:
+        raise Exception(f"Value for backend is undefined. Current value: {result!r}")
```
Comment (cpcloud) [inline ibis/backends/tests/test_generic.py:1174]: Can this be `xfail`d or `skip`ped instead?

File: ibis/backends/tests/test_generic.py [hunk 2]
```diff
@@ -760,9 +769,406 @@ def test_exists(batting, awards_players, method_name):
...
+        "datafusion",
+        "mssql",
+        "mysql",
+        "pandas",
+        "polars",
+        "pyspark",
+        "trino",
+    ],
+    raises=(NotImplementedError, com.OperationNotDefinedError),
+    strict=True,
```
Comment (cpcloud) [inline ibis/backends/tests/test_generic.py:1164]: This is already `True` by default, is it required here?

**Review by cpcloud — COMMENTED** [review]
Nice work!

File: ibis/backends/tests/test_generic.py (hunk 1 above)
Comment (krzysztof-kwitt) [reply ibis/backends/tests/test_generic.py:1174]: No. This is not a backend bug but a bug in the test and we should just add the missing value to the test. 

Before, I had code like below:
```
assert result == expected_values[backend_name]
```
But that didn't display the expected value, and it threw a KeyError exception. Now that I have the expected value, I can update the tests based on the CI results and I don't have to configure all environments.

Before merging, we can bring back the original form.

File: ibis/backends/tests/test_generic.py [hunk 3]
```diff
@@ -760,9 +769,406 @@ def test_exists(batting, awards_players, method_name):
...
+            {
+                "bigquery": "ARRAY",
+                'snowflake': "ARRAY",
+                "duckdb": "DOUBLE[]",
+                "postgres": "numeric[]",
+            },
+            marks=[
+                pytest.mark.broken(
+                    ['sqlite'],
+                    "(sqlite3.InterfaceError) Error binding parameter 0 - probably unsupported type.",
```
[the remark this answers is not in the document]
Comment (krzysztof-kwitt) [reply ibis/backends/tests/test_generic.py:1124]: I agree. We should throw an exception that this is an unsupported operation for SQLIte, but in this PR, I just want to cover the current behavior and then we can improve it.

File: ibis/backends/tests/test_generic.py [hunk 4]
```diff
@@ -760,9 +769,406 @@ def test_exists(batting, awards_players, method_name):
...
+                "duckdb": "DECIMAL(18,3)",
+                'sqlite': "real",
+                "postgres": "numeric",
+            },
+            id="decimal",
+            marks=[
+                pytest.mark.broken(
+                    ['bigquery'],
+                    "ufunc 'isfinite' not supported for the input types, and the inputs could not be "
+                    "safely coerced to any supported types according to the casting rule ''safe''",
```
[the remark this answers is not in the document]
Comment (krzysztof-kwitt) [reply ibis/backends/tests/test_generic.py:1108]: Yes. It is caused by this line.
https://github.com/ibis-project/ibis/blob/e4bae26b6671a16ceeab9beff4e21d66a6e2fbb9/ibis/backends/bigquery/registry.py#L221

File: ibis/backends/tests/test_generic.py [hunk 5]
```diff
@@ -760,9 +769,406 @@ def test_exists(batting, awards_players, method_name):
...
+                'snowflake': "TIMESTAMP",
+                "duckdb": "TIMESTAMP",
+                'sqlite': "text",
+                "postgres": "timestamp without time zone",
+            },
+            marks=[
+                pytest.mark.broken(
+                    ['snowflake'],
+                    '(snowflake.connector.errors.ProgrammingError) '
+                    '002140 (42601): SQL compilation error: Unknown functions NOW, TIMEZONE',
```
[the remark this answers is not in the document]
Comment (krzysztof-kwitt) [reply ibis/backends/tests/test_generic.py:1009]: snowflake has `current_timestamp` function.

[the remark this answers is not in the document]
Comment (cpcloud) [reply ibis/backends/tests/test_generic.py:1009]: Addressed in https://github.com/ibis-project/ibis/pull/5264.

**Comment by krzysztof-kwitt** [comment]
@cpcloud it is ready for review, but I don't know why the tests for Polars fail. Do you have an idea?

**Review by cpcloud — APPROVED** [review]
Looks there were some broken things in Polars that you discovered!

I fixed them up, this should be good to go now. Thank you!

**Comment by cpcloud** [comment]
BigQuery tests look good, merging:

```
…/ibis on  bigquery-typeof is 📦 v4.0.0 via 🐍 v3.10.9 via ❄️  impure (ibis-3.10.9)
❯ pytest -m bigquery -n auto --dist=loadgroup
======================================================================================== test session starts =========================================================================================
platform linux -- Python 3.10.9, pytest-7.2.1, pluggy-1.0.0
benchmark: 4.0.0 (defaults: timer=time.perf_counter disable_gc=False min_rounds=5 min_time=0.000005 max_time=1.0 calibration_precision=10 warmup=False warmup_iterations=100000)
Using --randomly-seed=2174117816
rootdir: /home/cloud/src/ibis, configfile: pyproject.toml
plugins: mock-3.10.0, repeat-0.9.1, benchmark-4.0.0, xdist-3.1.0, hypothesis-6.62.1, snapshot-0.9.0, clarity-1.0.1, profiling-1.7.0, randomly-3.12.0, cov-4.0.0
gw0 ok / gw1 [1072] / gw2 [1072] / gw3 [1072] / gw4 [1072] / gw5 [1072] / gw6 [1072] / gw7 [1072] / gw8 [1072] / gw9 [1072] / gw10 ok / gw11 ok / gw12 [1072] / gw13 [1072] / gw14 [1072] / gw15 [1072gw0 ok / gw1 [1072] / gw2 [1072] / gw3 [1072] / gw4 [1072] / gw5 [1072] / gw6 [1072] / gw7 [1072] / gw8 [1072] / gw9 [1072] / gw10 ok / gw11 [1072] / gw12 [1072] / gw13 [1072] / gw14 [1072] / gw15 [gw0 [1072] / gw1 [1072] / gw2 [1072] / gw3 [1072] / gw4 [1072] / gw5 [1072] / gw6 [1072] / gw7 [1072] / gw8 [1072] / gw9 [1072] / gw10 ok / gw11 [1072] / gw12 [1072] / gw13 [1072] / gw14 [1072] / gwgw0 [1072] / gw1 [1072] / gw2 [1072] / gw3 [1072] / gw4 [1072] / gw5 [1072] / gw6 [1072] / gw7 [1072] / gw8 [1072] / gw9 [1072] / gw10 [1072] / gw11 [1072] / gw12 [1072] / gw13 [1072] / gw14 [1072] / gw15 [1072]
xxx.....xx....x..xxx..xx..xx.x...x.xx...x.....xxxx.xx.........x.....x..x.xxx.x.....x.....x.....x...x....xxx..x...x.x.....x..x.x.x....x....x...x.x...xx....x......x.xx....................x.... [ 17%]
.....x..x.xx..x.x.....x..x.xx.....x........x.x........x..x.xx.xx...........xx.....x..x.x.x..x...xx..xx..x.x.x.s.....xx..x..xx.x..........x.....x.............x...x...x.....x...x.....x.x....x. [ 35%]
....x.x.x...x...x..x......x.xx................x.................x........xxxx.xxxxx.xxxx.xxx.x.xxxxxx..x.xxxxxxx.xxx.xxxxx.x..x..xx.x....xxx....x...xx...x......x..x.xxx.....x.......x.......x [ 53%]
.x..x.....x..x.........xx........x....x......x.x.......x...x..x........xx...........x......x.x...................x......x..............................x..........x..............x..x.x..x.... [ 70%]
.....................x........................x..............................x..........s.......x............................................................................................. [ 88%]
.................................................................................x...xxx.xx.......x.x.......xx............                                                                     [100%]Exception ignored in: <ssl.SSLSocket fd=15, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_STREAM, proto=6, laddr=('2600:4041:5f65:8200:1ac0:4dff:fe00:f1b', 45938, 0, 0), raddr=('2607:f8b0:4006:80d::200a', 443, 0, 0)>
ResourceWarning: unclosed <ssl.SSLSocket fd=15, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_STREAM, proto=6, laddr=('2600:4041:5f65:8200:1ac0:4dff:fe00:f1b', 45938, 0, 0), raddr=('2607:f8b0:4006:80d::200a', 443, 0, 0)>

====================================================================== 855 passed, 2 skipped, 215 xfailed in 110.81s (0:01:50) =======================================================================
```

**Pull request merged**